pyfr/backends/cuda/cublaslt.py

```diff
@@ -0,0 +1,296 @@
+            raise ValueError('Incompatible matrices for out = a*b')
+
+        # cuBLAS expects inputs to be column-major (or Fortran order in
+        # numpy parlance).  However as C = A*B => C^T = (A*B)^T
+        # = (B^T)*(A^T) with a little trickery we can multiply our
+        # row-major matrices directly.
+        m, n, k = b.ncol, a.nrow, a.ncol
+        A, B, C = b, a, out
+
+        # Cache key
+        ckey = (A.dtype, alpha, beta, m, n, k, A.leaddim, B.leaddim, C.leaddim)
+
+        # Size checks
+        if any(sz > 2**31 - 1 for sz in ckey[3:]):
+            raise ValueError('Matrices too large for cuBLAS')
+
+        # α and β factors for D = α*(A*B) + β*C
+        if a.dtype == np.float64:
+            alpha_ct, beta_ct = c_double(alpha), c_double(beta)
+            ctype = w.CUBLAS_COMPUTE_64F
+            dtype = w.CUDA_R_64F
+        else:
+            alpha_ct, beta_ct = c_float(alpha), c_float(beta)
+            ctype = w.CUBLAS_COMPUTE_32F
+            dtype = w.CUDA_R_32F
+
+        desc = GEMMDesc(None, None, None, None, None, None, 0)
+
+        # Convenience wrapper
+        def gemm(stream):
+            w.cublasLtMatmul(
+                h, desc.mm, byref(alpha_ct), A, desc.a, B,
+                desc.b, byref(beta_ct), C, desc.c, C, desc.c,
+                byref(desc.algo), desc.ws_ptr, desc.ws_size, stream
+            )
+
+        # Obtain the performance of the kernel
+        try:
+            desc, dt = self._mul_cache[ckey]
+        except KeyError:
+            # Create matrix layouts
+            a_desc = CUBLASLtMatrixLayout(self, h, A, dtype)
+            b_desc = CUBLASLtMatrixLayout(self, h, B, dtype)
+            c_desc = CUBLASLtMatrixLayout(self, h, C, dtype)
+
+            # Matmul descriptor
+            mm_desc = CUBLASLtMatmulDesc(self, h, ctype, dtype)
+            mm_desc[w.CUBLASLT_MATMUL_DESC_TRANSA] = w.CUBLAS_OP_N
+            mm_desc[w.CUBLASLT_MATMUL_DESC_TRANSB] = w.CUBLAS_OP_N
+            mm_desc[w.CUBLASLT_MATMUL_DESC_TRANSC] = w.CUBLAS_OP_N
+
+            # Heuristic preference descriptor
+            pref = CUBLASLtMatmulPreference(self, h)
+            pref[w.CUBLASLT_MATMUL_PREF_MAX_WORKSPACE_BYTES] = 1024 * 1024 * 4
+
+            # Get matmul heuristics
+            heurs = (CUBLASLtMatmulHeuristicResult * self.nkerns)()
+            nreturn = c_int()
+            w.cublasLtMatmulAlgoGetHeuristic(
+                h, mm_desc, a_desc, b_desc, c_desc, c_desc, pref, self.nkerns,
+                heurs, byref(nreturn)
+            )
+
+            # Save a copy of the contents of the output matrix
+            out_np = getattr(out, 'parent', out).get()
+
+            # Benchmark the kernel
+            best_kern = None
+            for heur in heurs[:nreturn.value]:
+                # Allocate gemm workspace
+                ws_size = heur.workspaceSize
+                ws = CUDADevAlloc(cuda, ws_size) if ws_size else None
```

Comment:
Allocate whatever the max workspace is up front in the constructor and pass this.

Reply:
The heuristic returns the workspace size needed, but sometimes it's zeros and this handles that.

Reply:
That's okay, but for benchmarking if you've already pre-allocated the maximum possible size then run with it.  Avoids lots of little memory allocations and deallocations.  Once you know the best kernel you can allocate precisely the right amount of workspace everytime someone requests the kernel.

Reply:
Gotcha, will do